### ISSUE TITLE
"Not found VSS volume header."

### ISSUE THREAD

**Issue by BeanBagKing** [mnrkbys/vss_carver#15]
I'm trying to do some research and documentation on this tool, but I'm continuously coming up with "Not found VSS volume header."

My methodology boils down to

1. Take a clean VM (Server 2019, although I've tried 2012 as well), VMWare Workstation Pro 16
2. Create a file
3. Enable VSS, first shadow automatically gets created
4. Verify vss is there with vssadmin list shadows
5. Delete the file w/sdelete (don't want it showing up in slack space)
6. Delete the VSS using the commands below this list.
7. Verify the vss is deleted with vssadmin list shadows
8. Shut down the VM

After this it gets into the regular forensics portion. I use qemu to convert the vmdk to a raw disk, mostly so I can get the offset easier. I've tried directly to the vmdk as well though. fdisk -l <filename.raw> to get the start and end, multiply the start of my windows drive slot by the sector bytes (512) to get the offset. offset, file name, and locations for the catalog and store go into vss_carver.py, which seems to run fine. After multiple tests though, it never finds the VSS volume header.

I'm not sure there is a fix here, I'm mostly looking into any insight as to why this is failing. Not giving the VSS enough time to create? Thin provisioned virtual disks? virtual disks residing on a solid state drive? Can you give me more information about what it means to not find the VSS volume header, and why that might be? Is Server 2019 shadows known to be supported? I can think of a ton of stuff that could go wrong, not all of it I can account for in my lab.

Deleting the vss
```
vssadmin delete shadows /all /quiet
wmic shadowcopy delete
wbadmin delete catalog -quiet
```

```
vss_carver 20200312
==================================================
Stage 1: Checking if VSS is enabled.
Volume size: 0xeea000000
Not found VSS volume header.
```

Edit: I did a much more in-depth explanation of my setup here: https://nullsec.us/carving-for/

**Comment by mnrkbys** [comment]
Extract the data at "offset + 0x1e00" of the raw image you are analyzing with hexdump, etc.
Then, compare the data with the byte array below:

```python
vss_identifier = b'\x6B\x87\x08\x38\x76\xC1\x48\x4E\xB7\xAE\x04\x04\x6E\x6C\xC7\x52'
```

If they do not match, it means that your disk image does not have Volume Snapshot or disabled it.
Unfortunately, I have never checked a disk image converted with qemu-img command.

**Comment by BeanBagKing** [comment]
Got to do some testing tonight, that offset, plus the pattern, helped a lot. Still some testing to do, but I have some theories now. TL;DR, I think it's VMware snapshots.

For those coming behind me, I did things a little backwars from the suggestion. Instead of extracting the data and looking at the hex at that location. I took a new server and grepped for the pattern.

```
LANG=C grep --only-matching --byte-offset --binary --text --perl-regexp "\x6B\x87\x08\x38\x76\xC1\x48\x4E\xB7\xAE\x04\x04\x6E\x6C\xC7\x52" smallserver.raw
```

This should return the byte offset of my pattern. In the case of my new server this is the sector start times 512 (the original vss_carver location) + 7680 (the 0x1e00 given above).

New server, only one pattern match, and at the wrong offset. Just to make sure, I took that offset, subtracted 7680 to get the location vss_carver would need to start at, and ran it. It found an "offset" (or what it through was one) started to process it, and promptly crashed, no surprise.

Enabled VSS, shut the server down, grepped again, and got 89 matches. One of these was at my exact sector*512+7680 location. Ran vss_carver, and it found the VSS.

Started the server back up, deleted the VSS, shutdown, grep, and got 85 matches. Something was deleted, but we still have some data. Ran vss_carver again and got the magical "VSS snapshot was enabled. But all snapshots were deleted."!

The big difference was that I never took a snapshot of this system. I took one of the previous system, to make rolling back the lab easier. I'm not a VMware guru, but just based on watching files, it looks like VMware stores incremental differences in a <name>-000001.vmdk file. This was kind of my guess when I tried rolling all the VMDK's together with vmware-vdiskmanager. I guess this isn't the proper way to get back to the "live" disk though.

@mnrkbys - feel free to close this. I think what you gave me was the key. Thank you again. If you want to leave it open though, I'll continue adding comments as I figure out how to recover from a "production" environment with numerous snapshots.
